File: Sources/CCUsageCore/Engine/SessionLauncher.swift
import Foundation

/// Prepara o lançamento de uma sessão num grupo: escolhe a conta certa, ativa-a,
/// e devolve o que o processo precisa para subir. O `exec` em si fica na CLI —
/// aqui é só a decisão, para ser testável sem lançar nada.
public struct SessionLauncher: Sendable {
    /// O que a CLI precisa para subir a sessão no grupo.
    public struct LaunchPlan: Sendable, Equatable {
        /// O executável do provedor (`claude`).
        public let executable: String
        /// Argumentos do lançamento (os que o usuário passou depois do grupo).
        public let arguments: [String]
        /// O valor de `CLAUDE_CONFIG_DIR`, ou `nil` no grupo padrão (que não
        /// exporta a variável — setá-la sobe a sessão deslogada).
        public let configDirEnv: String?
        /// A conta que ficou ativa para esta sessão.
        public let account: Account
    }

    public enum LaunchError: Error, Equatable {
        case unknownGroup(String)
        case emptyGroup(String)
        case noUsableAccount(String)
    }

    private let engine: RotationEngine
    private let adapter: any ProviderAdapter

    public init(keychain: any KeychainStore,
                adapters: [any ProviderAdapter] = [AnthropicAdapter()]) {
        self.engine = RotationEngine(keychain: keychain, adapters: adapters)
        self.adapter = adapters.first ?? AnthropicAdapter()
    }

    /// Acha um grupo pelo nome, sem diferenciar maiúsculas nem espaços nas pontas.
    public func group(named name: String, in config: RouterConfig) -> AccountGroup? {
        let wanted = name.lowercased().trimmingCharacters(in: .whitespaces)
        return config.groups.first {
            $0.name.lowercased().trimmingCharacters(in: .whitespaces) == wanted
        }
    }

    /// Escolhe a conta que vai servir a sessão e a ativa.
    ///
    /// Ordem da decisão: mantém a conta ativa se ela ainda tem folga; senão vai
    /// para a próxima com folga na ordem de preferência; senão, para a primeira
    /// que ao menos tenha credencial — é melhor subir numa conta cheia do que
    /// recusar o lançamento (o usuário pode trocar de modelo, ou só ler).
    public func prepare(group: AccountGroup, config: RouterConfig,
                        usage: [UUID: Double], arguments: [String]) throws -> LaunchPlan {
        let accounts = config.accounts(in: group)
        guard !accounts.isEmpty else { throw LaunchError.emptyGroup(group.name) }

        let active = engine.activeAccount(in: group, config: config)
        let threshold = group.thresholdPercent / 100

        let chosen: Account
        if let active, let used = usage[active.id], used < threshold {
            chosen = active  // ativa ainda tem folga
        } else if let next = engine.nextAccount(for: group, config: config, usage: usage) {
            chosen = next    // a melhor com folga
        } else if let active {
            chosen = active  // ninguém com folga; fica na atual
        } else {
            chosen = accounts.first!  // nunca ativou nenhuma; começa pela primeira
        }

        do {
            try engine.activate(chosen, in: group, config: config)
        } catch {
            throw LaunchError.noUsableAccount(group.name)
        }

        return LaunchPlan(
            executable: adapter.launchCommand().executable,
            arguments: arguments,
            configDirEnv: group.configDir.environmentValue,
            account: chosen)
    }
}

File: windows/crates/router-core/src/engine/session_launcher.rs
//! Prepara o lançamento de uma sessão num grupo (≙ `SessionLauncher.swift`):
//! escolhe a conta certa, ativa-a e devolve o que o processo precisa para subir.
//! O processo em si fica na CLI — aqui é só a decisão, testável sem lançar nada.

use std::collections::HashMap;
use std::sync::Arc;

use super::account_model::Account;
use super::anthropic_adapter::AnthropicAdapter;
use super::credential_store::CredentialStore;
use super::group_model::{AccountGroup, RouterConfig};
use super::provider::ProviderAdapter;
use super::rotation_engine::RotationEngine;
use crate::ids::Id;

/// O que a CLI precisa para subir a sessão no grupo.
#[derive(Clone, PartialEq, Debug)]
pub struct LaunchPlan {
    /// O executável do provedor (`claude`).
    pub executable: String,
    /// Os argumentos que o usuário passou depois do grupo.
    pub arguments: Vec<String>,
    /// O valor de `CLAUDE_CONFIG_DIR`, ou `None` no grupo padrão (que não exporta
    /// a variável — setá-la sobe a sessão deslogada).
    pub config_dir_env: Option<String>,
    /// A conta que ficou ativa para esta sessão.
    pub account: Account,
}

#[derive(Clone, PartialEq, Eq, Debug, thiserror::Error)]
pub enum LaunchError {
    #[error("grupo desconhecido: {0}")]
    UnknownGroup(String),
    #[error("o grupo {0} não tem contas — adicione uma no app")]
    EmptyGroup(String),
    #[error("nenhuma conta do grupo {0} pôde ser ativada — use Relogar no app")]
    NoUsableAccount(String),
}

pub struct SessionLauncher {
    engine: RotationEngine,
    adapter: Arc<dyn ProviderAdapter>,
}

impl SessionLauncher {
    pub fn new(
        credentials: Arc<dyn CredentialStore>,
        adapters: Vec<Arc<dyn ProviderAdapter>>,
    ) -> Self {
        let adapter = adapters
            .first()
            .cloned()
            .unwrap_or_else(|| Arc::new(AnthropicAdapter));
        SessionLauncher {
            engine: RotationEngine::new(credentials, adapters),
            adapter,
        }
    }

    pub fn engine(&self) -> &RotationEngine {
        &self.engine
    }

    /// Acha um grupo pelo nome, sem diferenciar maiúsculas nem espaços nas pontas.
    pub fn group_named<'c>(name: &str, config: &'c RouterConfig) -> Option<&'c AccountGroup> {
        let wanted = name.trim().to_lowercase();
        config
            .groups
            .iter()
            .find(|g| g.name.trim().to_lowercase() == wanted)
    }

    /// Escolhe a conta que vai servir a sessão e a ativa.
    ///
    /// Ordem: mantém a ativa se ela ainda tem folga; senão a próxima com folga na
    /// ordem de preferência; senão a ativa mesmo; senão (grupo nunca ativado) a
    /// primeira — é melhor subir numa conta cheia do que recusar o lançamento.
    pub fn prepare(
        &self,
        group: &AccountGroup,
        config: &RouterConfig,
        usage: &HashMap<Id, f64>,
        arguments: Vec<String>,
    ) -> Result<LaunchPlan, LaunchError> {
        let accounts = config.accounts_in(group);
        let Some(first) = accounts.first() else {
            return Err(LaunchError::EmptyGroup(group.name.clone()));
        };
        let active = self.engine.active_account(group, config);
        let threshold = group.threshold_percent / 100.0;

        let chosen = match active {
            Some(a) if usage.get(&a.id).is_some_and(|u| *u < threshold) => a,
            _ => match self.engine.next_account(group, config, usage) {
                Some(next) => next,
                None => active.unwrap_or(first),
            },
        };

        self.engine
            .activate(chosen, group, config)
            .map_err(|_| LaunchError::NoUsableAccount(group.name.clone()))?;

        Ok(LaunchPlan {
            executable: self.adapter.launch_command().0,
            arguments,
            config_dir_env: group.config_dir.environment_value().map(String::from),
            account: chosen.clone(),
        })
    }
}
